EIPS/eip-2124.md

```diff
@@ -0,0 +1,294 @@
+    - The two nodes are in the same fork state currently. They might know of differing future forks, but that's not relevant until the fork triggers (might be postponed, nodes might be updated to match).
+2) If the remote `FORK_HASH` is a subset of the local past forks and the remote `FORK_NEXT` matches with the locally following fork block number, connect.
+    - Remote node is currently syncing. It might eventually diverge from us, but at this current point in time we don't have enough information.
+3) If the remote `FORK_HASH` is a superset of the local past forks and can be completed with locally known future forks, connect.
+    - Local node is currently syncing. It might eventually diverge from the remote, but at this current point in time we don't have enough information.
+4) Reject in all other cases.
+
+#### Stale software examples
+
+The examples below try to exhaust the fork combination possibilities that arise when nodes do not run matching software versions, but otherwise follow the same chain (mainnet nodes, testnet nodes, etc).
+
+| Past forks | Future forks | Remote `FORK_HASH` | Remote `FORK_NEXT` | Connect | Reason |
+|:---:|:---:|:---:|:---:|:---:|:---:|
+| A |  | A |  | Yes (1) | Same forks, same sync state. |
+| A |  | A | B | Yes (1) | Remote is advertising a future fork, but that is uncertain. |
+| A | B | A |  | Yes (1) | Local knows about a future fork, but that is uncertain. |
+| A | B | A | B | Yes (1) | Both know about a future fork, but that is uncertain. |
+| A | B1 | A | B2 | Yes (1) | Both know about differing future forks, but those are uncertain. |
+| [A,B] |  | A | B | Yes (2) | Remote out of sync. |
+| [A,B,C] |  | A | B | Yes¹ (2) | Remote out of sync. Remote will need a software update, but we don't know it yet. |
+| A | B | A ⊕ B |  | Yes (3) | Local out of sync. |
+| A | B,C | A ⊕ B |  | Yes (3) | Local out of sync. Local also knows about a future fork, but that is uncertain yet. |
+| A |  | A ⊕ B |  | No (4) | Local needs software update. |
+| A | B | A ⊕ B ⊕ C | | No² (4) | Local needs software update. |
+| [A,B] |  | A |  | No (4) | Remote needs software update. |
+
+*Note, there's one asymmetry in the table, marked with ¹ and ². Since we don't have access to a remote node's future fork list (just the next one), we can't detect that it's software is stale until it syncs up. This is acceptable as 1) the remote node will disconnect from us anyway, and 2) this is a temporary fluke during sync, not permanent with a leftover node.*
+
+#### Mismatching chain examples (local perspective)
+
+TODO: Give some examples as to what happens if the nodes follow different forks.
+
+## Rationale
+
+##### Why flatten `FORK_HASH` into 4 bytes? Why not share the entire genesis and fork list?
+
+Whilst the `eth` devp2p protocol permits arbitrarily much data to be transmitted, the discovery protocol relies on MTU-limited UDP messages. The total space allowance for all ENR entries is 300 bytes.
+
```

Comment:
I would like to request removing the MTU thing. It's OK to just say the identifier should be small to fit into ENR. The MTU is not related at all.

Reply:
Removed.